cpp/src/arrow/array/array_nested.cc

```diff
@@ -847,14 +847,13 @@ Result<std::shared_ptr<Array>> MapArray::FromArraysInternal(
   const auto& typed_offsets = checked_cast<const OffsetArrayType&>(*offsets);
 
   BufferVector buffers;
-  int64_t null_count;
-  if (null_bitmap != nullptr) {
-    buffers = BufferVector({std::move(null_bitmap), typed_offsets.values()});
-    null_count = null_bitmap->size();
-  } else {
-    buffers = BufferVector({null_bitmap, typed_offsets.values()});
-    null_count = 0;
+  buffers.resize(2);
+  int64_t null_count = 0;
+  if (null_bitmap) {
+    buffers[0] = std::move(null_bitmap);
```

Comment:
Change `const std::shared_ptr<Buffer>& null_bitmap` to `std::shared_ptr<Buffer> null_bitmap` so the `std::move` is effective. That would mean callers that pass null_bitmap as an `&&` don't cause a single reference increment on the atomic refcount integer of the shared_ptr.

(I know this is not the style in the other `std::shared_ptr` parameters but is the direction we want to go with parameters in constructors of factory functions in general).

Reply:
Sorry you had to make it more explicit and repeat what you have already said in https://github.com/apache/arrow/pull/41757#discussion_r1624758057. I missed it and will correct, thanks!

Reply:
@felipecrv for my education, is there a reason why we would want to specifically pass `null_bitmap` as a plain shared_ptr, while the other data (offsets, keys, items arrays) are passed as const ref?

Reply:
Sorry, reading your comment above more carefully, you actually already explain it ;)



> (I know this is not the style in the other `std::shared_ptr` parameters but is the direction we want to go with parameters in constructors of factory functions in general).

Is this documented somewhere? For people not following the C++ development daily or less familiar with C++ (there are also a lot of different ways to write C++ code ..), it might be useful to have some guidelines about which typical patterns we want to use in Arrow (not sure if that would fit in https://arrow.apache.org/docs/dev/developers/cpp/conventions.html)

Reply:
The closest to documentation that I can find to cite: https://isocpp.github.io/CppCoreGuidelines/CppCoreGuidelines#Rr-sharedptrparam-owner

`clang-tidy` also has a rule that complains about this (but note that we don't enforce clang-tidy rules in Arrow currently).

I can also explain the mechanics with an example.

This is how we currently have some classes (ctor and factories taking `const&`):

```cpp
class ListType {
  ListType(const std::shared_ptr<DataType>& type)
    : value_type_(type) {}
}

std::shared_ptr<ListType> list(const std::shared_ptr<DataType>& type) {
  return std::make_shared<ListType>(type);
}
```

When I run:

```cpp
auto value_type = int8();  // creation of a shared_ptr (ref-count = 1)
// value_type is passed as a reference to the shared_ptr
// (a pointer to the smart pointer instance)
auto list_type = list(value_type);
  // this in turn will call the ListType constructor by passing a reference
  ListType::ListType(value_type);
    // which will increment the ref-count of the shared_ptr on this line
    : value_type_(type) { // ref-count = 2

// at the end of the scope `list_type` is destroyed
// which calls the `value_type_` destructor, since ref-count = 2,
// it only decrements the ref-count which ends-up at 1

// now value_type is destroyed after the ref-count reaches 0,
// so `DataType::~DataType()` is called
```

If instead we have ctors and factories be like this:

```cpp
class ListType {
  ListType(std::shared_ptr<DataType> type)
    : value_type_(std::move(type)) {} // note the move
}

std::shared_ptr<ListType> list(std::shared_ptr<DataType> type) {
  return std::make_shared<ListType>(std::move(type));
}
```

when I run

```cpp
auto value_type = int8();  // creation of a shared_ptr (ref-count = 1)
// value_type is passed as a && to the shared_ptr because of the std::move
// (it's still a pointer, but it says that `value_type` can become "moved-from")
auto list_type = list(std::move(value_type));
  // this in turn will call the ListType constructor by passing a reference
  ListType::ListType(std::move(value_type));
    // which will construct value_type_ from an && meaning that
    // the ref-count stays the same (1) and `value_type` is now a moved-from `nullptr`
    : value_type_(std::move(type)) { // ref-count = 1

// at the end of the scope `list_type` is destroyed
// which calls the `value_type_` destructor, since ref-count = 1,
// it calls `DataType::~DataType()` and destroys the `Int8Type` instance

// now value_type is destroyed but since it's `nullptr` nothing
// needs to happen
```

Since `shared_ptr`'s ref-count is atomic, the overhead of atomically updating the integer adds up. The second version does 0 increments and 1 decrement while the first has to increment 1 time and perform 2 decrements.

Reply:
Thank you for such a detailed response and for taking time to educate us! ❤️

Reply:
I might be missing something but removing `const` in `null_bitmap ` from `FromArraysInternal` and both `MapArray::FromArrays` I get the following error:

```
/Users/alenkafrim/repos/arrow/cpp/src/arrow/array/array_nested.h:540:32: error: non-const lvalue reference to type 'std::shared_ptr<Buffer>' cannot bind to a temporary of type 'std::nullptr_t'
      std::shared_ptr<Buffer>& null_bitmap = NULLPTR);
                               ^             ~~~~~~~
/Users/alenkafrim/repos/arrow/cpp/src/arrow/array/array_nested.h:540:32: note: passing argument to parameter 'null_bitmap' here
/Users/alenkafrim/repos/arrow/cpp/src/arrow/array/array_nested.h:546:32: error: non-const lvalue reference to type 'std::shared_ptr<Buffer>' cannot bind to a temporary of type 'std::nullptr_t'
      std::shared_ptr<Buffer>& null_bitmap = NULLPTR);
                               ^             ~~~~~~~
/Users/alenkafrim/repos/arrow/cpp/src/arrow/array/array_nested.h:546:32: note: passing argument to parameter 'null_bitmap' here
/Users/alenkafrim/repos/arrow/cpp/src/arrow/array/array_nested.h:566:50: error: non-const lvalue reference to type 'std::shared_ptr<Buffer>' cannot bind to a temporary of type 'std::nullptr_t'
      MemoryPool* pool, std::shared_ptr<Buffer>& null_bitmap = NULLPTR);
```

the diff is:

```
diff --git a/cpp/src/arrow/array/array_nested.cc b/cpp/src/arrow/array/array_nested.cc
index fb2efe9ae..49f67a180 100644
--- a/cpp/src/arrow/array/array_nested.cc
+++ b/cpp/src/arrow/array/array_nested.cc
@@ -807,7 +807,7 @@ MapArray::MapArray(const std::shared_ptr<DataType>& type, int64_t length,
 Result<std::shared_ptr<Array>> MapArray::FromArraysInternal(
     std::shared_ptr<DataType> type, const std::shared_ptr<Array>& offsets,
     const std::shared_ptr<Array>& keys, const std::shared_ptr<Array>& items,
-    MemoryPool* pool, const std::shared_ptr<Buffer>& null_bitmap) {
+    MemoryPool* pool, std::shared_ptr<Buffer>& null_bitmap) {
   using offset_type = typename MapType::offset_type;
   using OffsetArrowType = typename CTypeTraits<offset_type>::ArrowType;
 
@@ -861,7 +861,7 @@ Result<std::shared_ptr<Array>> MapArray::FromArraysInternal(
 Result<std::shared_ptr<Array>> MapArray::FromArrays(
     const std::shared_ptr<Array>& offsets, const std::shared_ptr<Array>& keys,
     const std::shared_ptr<Array>& items, MemoryPool* pool,
-    const std::shared_ptr<Buffer>& null_bitmap) {
+    std::shared_ptr<Buffer>& null_bitmap) {
   return FromArraysInternal(std::make_shared<MapType>(keys->type(), items->type()),
                             offsets, keys, items, pool, null_bitmap);
 }
@@ -869,7 +869,7 @@ Result<std::shared_ptr<Array>> MapArray::FromArrays(
 Result<std::shared_ptr<Array>> MapArray::FromArrays(
     std::shared_ptr<DataType> type, const std::shared_ptr<Array>& offsets,
     const std::shared_ptr<Array>& keys, const std::shared_ptr<Array>& items,
-    MemoryPool* pool, const std::shared_ptr<Buffer>& null_bitmap) {
+    MemoryPool* pool, std::shared_ptr<Buffer>& null_bitmap) {
   if (type->id() != Type::MAP) {
     return Status::TypeError("Expected map type, got ", type->ToString());
   }
diff --git a/cpp/src/arrow/array/array_nested.h b/cpp/src/arrow/array/array_nested.h
index f96b6bd3b..173147f55 100644
--- a/cpp/src/arrow/array/array_nested.h
+++ b/cpp/src/arrow/array/array_nested.h
@@ -537,13 +537,13 @@ class ARROW_EXPORT MapArray : public ListArray {
   static Result<std::shared_ptr<Array>> FromArrays(
       const std::shared_ptr<Array>& offsets, const std::shared_ptr<Array>& keys,
       const std::shared_ptr<Array>& items, MemoryPool* pool = default_memory_pool(),
-      const std::shared_ptr<Buffer>& null_bitmap = NULLPTR);
+      std::shared_ptr<Buffer>& null_bitmap = NULLPTR);
 
   static Result<std::shared_ptr<Array>> FromArrays(
       std::shared_ptr<DataType> type, const std::shared_ptr<Array>& offsets,
       const std::shared_ptr<Array>& keys, const std::shared_ptr<Array>& items,
       MemoryPool* pool = default_memory_pool(),
-      const std::shared_ptr<Buffer>& null_bitmap = NULLPTR);
+      std::shared_ptr<Buffer>& null_bitmap = NULLPTR);
 
   const MapType* map_type() const { return map_type_; }
 
@@ -563,7 +563,7 @@ class ARROW_EXPORT MapArray : public ListArray {
   static Result<std::shared_ptr<Array>> FromArraysInternal(
       std::shared_ptr<DataType> type, const std::shared_ptr<Array>& offsets,
       const std::shared_ptr<Array>& keys, const std::shared_ptr<Array>& items,
-      MemoryPool* pool, const std::shared_ptr<Buffer>& null_bitmap = NULLPTR);
+      MemoryPool* pool, std::shared_ptr<Buffer>& null_bitmap = NULLPTR);
```

Reply:
Done: https://github.com/apache/arrow/pull/41956/commits/a8586cabb4749069a6b67c335fede996b1ed65eb